text/0000-eliminate-finalize.md

```diff
@@ -0,0 +1,142 @@
+* The sender's partial signature will be `sender.partial_sig = sender.nonce + e * sender.excess`
+* The receiver's partial signature will be `receiver.partial_sig = receiver.nonce + e * receiver.excess` 
+
+The final aggregated kernel signature `(s, k*G)` can then be calculated as `(sender.partial_sig + receiver.partial_sig + proof_nonce, total_nonce * G)`
+
+### Verifying the proof
+
+The following 4 steps are necessary for the sender to prove payment to the receiver's address:
+
+1. Provide the kernel and show that it was confirmed on-chain.
+2. Prove knowledge of `sender.nonce`
+3. Show that `sender.public_nonce + receiver.public_nonce + (proof_nonce * G)` = `total_nonce * G` (ie. the `k*G` in the kernel signature)
+4. Provide the preimage to `proof_nonce` (`diffie_secret | receiver.public_excess | tx.amount`)
+
+## Slate Format
+
+A slate is now only needed for passing information from the sender to receiver. It must include the following data:
+
+* `tx.amount`
+* `tx.offset`
+* `tx.kernel_features`
+* `tx.fee`
+* `tx.inputs`
+* `receiver.change_output`
+* `receiver.public_excess` & `receiver.public_nonce` (to prevent the need for grinding)
+* `sender.public_excess`, `sender.public_nonce`, & `sender.partial_sig`
+* `time_to_live` (optional)
+
+TODO: Define JSON & binary formats
+TODO: Mention QR codes
+
+## Backward Compatibility
+
+TODO: Decide how to handle versioning. In theory, this should be simple. If given an old slatepack address, use the old slate format. If given a new, one-time address, then the new slate format is supported.
+
+## Invoices
+
+To send an invoice to a payer, the recipient provides the payer with a newly generated one-time address, an amount, and any additional fields we decide to support (memo, refund address, expiration, etc). The payer, upon approving payment of the invoice, simply follows the standard 2-step send process described in this RFC. There's no need for a special workflow.
+
+## Security
+
+### Nonce Reuse
+
+Nonces should be randomly generated by the receiving wallet, rather than derived from seed, since reuse of a nonce could lead to loss of funds. Receiving wallets should keep track of all one-time addresses (excesses & nonces) they generate in a list (`address_list`), along with an indicator for whether they have been used yet.
+
+Upon receipt of a partial transaction (slate), the receiver should obtain a lock on `address_list` and then check whether the address has been used before. If it has not, it's safe to finalize the transaction, but before broadcasting, the address should be marked as used in the `address_list`, and then the lock can be safely released.
+
+### Play Attacks
+
+TODO: Describe play attacks, and finish this section.
+
+Any inputs sent from a wallet to a one-time address should be considered as sent. They should not be cancelable, however, the sender should have the option to add additional inputs and/or increase the fee (equivalent of bitcoin's RBF).
```

Comment:
Please correct me if I'm misunderstanding this, but I think that they should definitely be cancelable from the sender side. Consider a scenario where Alice is sending 5 Grin to Bob and she uses an input that holds 1000 Grin. She creates her change output with 995 Grin. If she can't cancel the transaction, then if Bob simply refuses to finalize the transaction, this is somewhat equivalent to Bob burning all the outputs which would make Alice lose her 995 Grin output. In theory, Bob could blackmail Alice telling her to send some more coins if she wants to get her change output. Adding a cancel option to Alice solves this as she could reuse the inputs in a new transaction.

Reply:
Good point. I didn't think that through very well at all 😆 
I guess I more meant that cancel and resend to same receiver because the transaction "failed" (a common case today) shouldn't be supported, since cancel requires respending those inputs. Instead, we should resend the original partial slate, perhaps with a larger fee. We should never naively just return the inputs back to the wallet as spendable.

Reply:
yeah, I agree. A 'resend' after cancelation should reuse the inputs to avoid the play attacks - a reuse of completely the same inputs and outputs is one such option.